doc/sphinx/proof-engine/tactics.rst

```diff
@@ -898,211 +1109,156 @@ Applying theorems
-
-   If the goal is a non-dependent product :math:`T \rightarrow U`, then it
-   puts in the local context either :g:`Hn:T` (if :g:`T` is of type :g:`Set`
-   or :g:`Prop`) or :g:`Xn:T` (if the type of :g:`T` is :g:`Type`).
-   The optional index ``n`` is such that ``Hn`` or ``Xn`` is a fresh
-   identifier. In both cases, the new subgoal is :g:`U`.
-
-   If the goal is an existential variable, :tacn:`intro` forces the resolution
-   of the existential variable into a dependent product :math:`\forall`\ :g:`x:?X, ?Y`,
-   puts :g:`x:?X` in the local context and leaves :g:`?Y` as a new subgoal
-   allowed to depend on :g:`x`.
+   .. exn:: @ident is already used.
+      :undocumented:
 
-   The tactic :tacn:`intro` applies the tactic :tacn:`hnf`
-   until :tacn:`intro` can be applied or the goal is not head-reducible.
+   .. exn:: No such hypothesis: @ident.
+      :undocumented:
 
    .. exn:: No product even after head-reduction.
       :undocumented:
 
-   .. tacv:: intro @ident
+   .. _intro_examples:
+   .. example:: `intro` and `intros`
 
-      This applies :tacn:`intro` but forces :token:`ident` to be the name of
-      the introduced hypothesis.
+      .. coqtop:: reset out
 
-      .. exn:: @ident is already used.
-         :undocumented:
+         Goal forall m n, m < n -> (let x := 0 in True).
 
-   .. note::
+      .. coqtop:: all
 
-      If a name used by intro hides the base name of a global constant then
-      the latter can still be referred to by a qualified name
-      (see :ref:`Qualified-names`).
+         intro m.
+         intro n.
+         intro H.
+         intro x.
 
-   .. tacv:: intros
-      :name: intros
+      This single `intros` tactic is equivalent to the 4 preceding `intro` tactics:
+
+      .. coqtop:: reset out
 
-      This repeats :tacn:`intro` until it meets the head-constant. It never
-      reduces head-constants and it never fails.
+         Goal forall m n, m < n -> (let x := 0 in True).
 
-   .. tacv:: intros {+ @ident}.
+      .. coqtop: all
 
-      This is equivalent to the composed tactic :n:`intro @ident; ... ; intro @ident`.
+         intros m n H x.
 
-   .. tacv:: intros until @ident
+.. tacn:: intros {* @intropattern }
+          intros until {| @ident | @natural }
 
-      This repeats intro until it meets a premise of the goal having the
-      form :n:`(@ident : @type)` and discharges the variable
-      named :token:`ident` of the current goal.
+      The first form introduces zero or more items into the context from the goal
+      using the :n:`intro` tactic.
+      If :n:`@intropattern` is not specified, the tactic repeats :tacn:`intro`
+      until it reaches the :term:`head constant`.  The :n:`@intropattern` is
+      often just a list of :n:`ident`\s, but other forms can also be specified
+      in order to, for example, introduce all quantified variables (`*`);
+      introduce all hypotheses before `->` and all quantified variables (`**`);
```

Comment:
```suggestion
      introduce all quantified variables and non-dependent premises (`**`);
```

Reply:
IIUC in this case, "quantified variables" doesn't include `exists`.  Not going to change it now, but the wording is inconsistent with mathematical usage.  But maybe later if I have a good idea.

Reply:
You are right! Then we shouldn't say quantified variables. And this form would be strictly more correct:
```suggestion
      introduce all dependent and non-dependent premises;
```